### PR TITLE
DP-719: Fix download of binaries

### PR DESCRIPTION
**Pull request by Bakrog**
This should fix issue #9

### CODE REVIEW

File: src/__tests__/paxfulapi_test.ts [hunk 1]
```diff
@@ -46,6 +50,40 @@ function mockCredentialsStorageReturnValue() {
...
+        }
+    }, {
+        status: 200,
+        body: JSON.stringify(expectedAnswer)
+    }, {
+        sendAsJson: false
+    });
+
+    const paxfulApi = usePaxful(credentials, credentialStorage);
+    const answer = await paxfulApi.invoke(paxfulTradeChatImageUploadUrl, image);
```
Comment (sergeil) [inline src/__tests__/paxfulapi_test.ts:82]: trade-chat/image/upload beside a binary also accepts "trade_hash" parameter, so we have endpoint which beside binary also accept some parameters

File: src/commands/Invoke.ts [hunk 2]
```diff
@@ -1,27 +1,37 @@
...
 // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
-const createRequest = (url: string, config: ApiConfiguration, credentials: Credentials, payload?: Record<string, unknown> | []): Request => {
+const createRequest = (url: string, config: ApiConfiguration, credentials: Credentials, payload?: InvokeBody): Request => {
     const headers = {
         "Accept": "application/json",
         "Content-Type": "application/x-www-form-urlencoded",
         "Authorization": `Bearer ${credentials.accessToken}`
     };
+    let body: BodyInit | undefined;
+    if (payload && (payload instanceof ReadStream || payload instanceof Buffer)) {
```
Comment (sergeil) [inline src/commands/Invoke.ts:22]: Again referring to trade-chat/image/upload, payload may be an object, whose one key "file" is binary, but "trade_hash" is a plain string. I guess here we need to recursively iterate payload's value to determine for sure what mime and how payload should be handled.

**Review by sergeil — CHANGES_REQUESTED** [review]

File: src/commands/Invoke.ts (hunk 2 above)
Comment (Bakrog) [reply src/commands/Invoke.ts]: No need to iterate to determinate the mimes, just need to use `FormData`.

File: src/__tests__/paxfulapi_test.ts (hunk 1 above)
Comment (Bakrog) [reply src/__tests__/paxfulapi_test.ts]: Changed the object.

File: src/commands/Invoke.ts [hunk 3]
```diff
@@ -1,30 +1,63 @@
...
 // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
-const createRequest = (url: string, config: ApiConfiguration, credentials: Credentials, payload?: Record<string, unknown> | []): Request => {
+const createRequest = (url: string, config: ApiConfiguration, credentials: Credentials, payload?: InvokeBody): Request => {
     const headers = {
         "Accept": "application/json",
         "Content-Type": "application/x-www-form-urlencoded",
         "Authorization": `Bearer ${credentials.accessToken}`
     };
+    let body: BodyInit | undefined;
+    if (url.endsWith('/trade-chat/image/upload') && payload) {
```
Comment (sergeil) [inline src/commands/Invoke.ts:23]: As a quick hack that will do, but we need to have more flexible solution. In other words, this method should iterate values of payload and if any of them is of type, say, ArrayBuffer, then we would switch to `multipart/form-data` content-type.

Comment (Bakrog) [reply src/commands/Invoke.ts:23]: Yes I tried to do something like that but the amount of code + how complex it had become it doesn’t feel like paying for a just more simple implementation right now as we just have one endpoint that is `multipart/form-data`. If in future we had a new endpoint we can interact over this code again with new tests and improve the code accordingly. What do you think?

Comment (sergeil) [reply src/commands/Invoke.ts:23]: Yep, for now will do, let's return to this topic later.

**Review by sergeil — APPROVED** [review]

**Pull request merged**